Python/Product/Analysis/LanguageServer/ILanguageServerExtensionProvider.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    /// <summary>
+    /// Implemented on a class that can create a language server extension.
+    /// This class must have a default constructor.
+    /// </summary>
+    public interface ILanguageServerExtensionProvider {
+        /// <summary>
+        /// Called when the extension is loaded for a language server.
+        /// </summary>
+        ILanguageServerExtension Create(IServer server, IReadOnlyDictionary<string, object> properties);
```

Comment:
I think this one needs to be async with cancellation token so we can abandon extension creation (i.e. it won't  be able to hang us)

Reply:
We're async all the way down, so may as well. But it's unlikely to be an issue - extensions at this stage shouldn't have to do more than object instantiation and event subscription.